cpp/pgen2.cc

```diff
@@ -2,14 +2,21 @@
 
 #include "pgen2.h"
 
-namespace parse {
+#include <vector>
 
-void Parser::setup(int start) {
-  assert(0);
-}
+namespace pnode {
+
+static std::vector<PNode> arena;
 
-bool Parser::addtoken(int typ, syntax_asdl::Token* opaque, int ilabel) {
-  assert(0);
+PNode* NewPNode(int typ, syntax_asdl::Token* tok,
+                List<pnode::PNode*>* children) {
+  if (tok == nullptr) {
```

Comment:
One thing occurred to me -- I think it would be better to add a Python stub in `Parser::setup()` like `pgen2::InitArena()` or something.   That way it's explicit, and we're not mixing the logic with allocation info

Though another thing I thought of is that we should make `_ParseOil` free the memory

You can imagine a test like this at the INTERACTIVE shell:

```
$ var x = 42 + 1
$ var x = 42 + 1
...
... # do this a million times
```

It shouldn't leak PNodes, they can be cleaned up every time

One way to do this might be to leverage the `ctx` pattern, like

```
with ctx_PNode(...):
  e_parser.Parse(...)
```

But we could do this in a separate step, since it could be tricky

It doesn't have  to be done all at once

Reply:
Ah yes, pulling this out of `NewPNode()` would be much easier to understand.

We can probably pull off the freeing thing you mentioned by tracking the pnode allocator as a member of `Parser`. It looks like `ExprParser` calls `Parser::setup()` at the beginning of every call to `ExprParser::Parse()`. I'll try that out